modules/cudacodec/include/opencv2/cudacodec.hpp

```diff
@@ -334,16 +370,35 @@ class CV_EXPORTS_W RawVideoSource
 
     /** @brief Updates the coded width and height inside format.
     */
-    virtual void updateFormat(const int codedWidth, const int codedHeight) = 0;
+    virtual void updateFormat(const FormatInfo& videoFormat) = 0;
+
+    /** @brief Signals RawVideoSource to start writing the raw bitstream to the given file from the next key frame.
+
+    @param filename Full path to the output file.  Each call to this with a new filename, filenameNew will wait for
+    the next key frame, stop writing to filename and start writing to filenameNew.  If VideoReader is already writing
+    to a file and filenameNew is empty, writing to filename will stop imidiately.
+    @param autoDetectExt Automatically detect and append the codec extension to filename before writing.
+
+    The function is intended to be used when streaming from an RTSP source where it is either
+    a) desirable to store the original streamed data, or
+    b) the overhead of re-encoding the decoded frame is undesirable.
+
+    @note This should only be used when streaming from rtsp and is not guaranteed to work when reading from a
+    file, especially from container formats avi, mp4 etc.  If the filename provided is invalid, cannot be opened
+    or written to, the first call to nextFrame() after calling this function will return false.
+     */
+    virtual void writeToFile(const std::string filename, const bool autoDetectExt = false) = 0;
 };
 
 /** @brief Creates video reader.
 
 @param filename Name of the input video file.
+@param filenameToWrite Full path to the raw output file.  If empty the raw encoded video data will not be written.
+@param autoDetectExt Automatically detect and append the codec extension to filename.
 
 FFMPEG is used to read videos. User can implement own demultiplexing with cudacodec::RawVideoSource
  */
-CV_EXPORTS_W Ptr<VideoReader> createVideoReader(const String& filename);
+CV_EXPORTS_W Ptr<VideoReader> createVideoReader(const String& filename, const String filenameToWrite = "", const bool autoDetectExt = false);
```

Comment:
> Main problem of this approach is missing unicode support.

Ahh I see.  How can we support unicode in the new api (`VideoReader_W`) for `const String& filename`?

> add extra new API with std::ostream&& or std::ofstream&& (note, move ctor is available only)

If possible I would like to be able to automatically detect the extension before creating the file so that the file can be played back without having to be renamed after writing has finished.

If this is not performed internally then this would involve creating a VideoCapture object before creating the new class just to detect the codec and then using that info to determine the file extension.
```
VideoCapture cap("rtsp://...")
std::string ext = FourCCToExtension(cap.get(CAP_PROP_FOURCC))
std::ofstream fileOut = (fileName + ext);
cv::Ptr<cv::cudacodec::VideoReader_W> reader_w = cv::cudacodec::createVideoReader_W(inputFile, fileOut);
```
I would like to avoid that can you suggest and alternative?

Reply:
> How can we support unicode in the new api (VideoReader_W) for const String& filename

`String` may handle UTF-8 strings only (it is std::string , BTW). There is no `wchar_t` or `std::wstring` in OpenCV.
Main question is about support on target platform (e.g., support of UTF-8 in `std::ofstream`) - Win 10 brings some support for UTF-8 but it should be pre-configured (also this configuration may break legacy apps).

---

> alternative

Capture to temporary file and then `rename()` with correct extension on user side.
But it also looks weird, so I don't have strong suggestion or recommendation here.